### PR TITLE
Fixes root password set up, errors when adding or updating users, and updated documentation

### LINKED ISSUE #214: Password management in mariadb_server_install is not working
When running the resource mariadd_server_install for the first time... A password is generated. But then, it's not saved. So at the next run, when it tries to re-generate one... And don't have the old one, convergence fail.

To be able to use this feature... Either we choose to create a .my.cnf, either we disable it. Don't see any easy workaround, as it's not like postgresql.

### PR DESCRIPTION
**Pull request by t0rrant**
## Description

This started as a fix for the issues below, but quickly escalated to other issues.

This PR mainly fixes setting the root password in the server_install recipe. The way this is achieved is by first killing off any mysql service that may be running (and which are properly registered in /var/run/mysql/mysqld.pid), launching up a mysqld service manually, with an initial script which sets the root password, killing it afterwards. Then it tries to authenticate with that password as a verification step.

Properly checks for the `node['mariadb']['server_root_password']` attribute being set, as well as the `password` property being passed. 

Fixes errors about the way the `only_if` guard is evaluated, it is now evaluated as a code block.

Also removed a reference in the integration tests about the `reload` function for mariadb service, which does not support this function.

Added property `apt_key_proxy` to `mariadb_repository` resource which allows passing an http_proxy to the apt-key command.

Fixed some mismatching quotes which made the `database_has_password_column` fail.

### Issues Resolved

#214 #220 

### Check List
- [x] All tests pass. See https://github.com/sous-chefs/apache2/blob/master/TESTING.md
  tested with kitchen in Ubuntu 16.04, Ubuntu 18.04, Debian 9 and CentOS 7
- [x] New functionality includes testing.
  the testing is the recipe not failing by itself
  for the new feature, I don't have any proxy available for testing

### CODE REVIEW

**Comment by t0rrant** [comment]
Hopefully I've got it right this time, at least from my end I can't see anything failing or with an unexpected behaviour, any tips on improving the solution are appreciated

**Comment by kitchen-porter** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please include a <a href="https://github.com/sous-chefs/line-cookbook/blob/master/CHANGELOG.md">CHANGELOG</a> entry.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by t0rrant** [comment]
btw, sorry about the mess with the open/close of PR's, it's my first time doing so many iterations on github eventually realized I didn't need to close them, just create new commits.

**Comment by t0rrant** [comment]
@Xorima can you review this, or request to be reviewed?

Cheers

File: metadata.rb [hunk 1]
```diff
@@ -7,7 +7,7 @@
 source_url       'https://github.com/sous-chefs/mariadb'
 issues_url       'https://github.com/sous-chefs/mariadb/issues'
 chef_version     '>= 13'
-version          '2.0.0'
+version          '2.0.3'
```
Comment (xorima) [inline metadata.rb]: Please revert, we prefer to control the version numbers ourselves :)

File: resources/server_install.rb [hunk 2]
```diff
@@ -24,7 +24,7 @@
 property :extconf_directory, String,        default: lazy { ext_conf_dir }
 property :data_directory,    String,        default: lazy { data_dir }
 property :external_pid_file, String,        default: lazy { "/var/run/mysql/#{version}-main.pid" }
-property :password,          [String, nil], default: 'generate'
+property :password,          [String, nil], default: lazy { node['mariadb']['server_root_password'] }
```
Comment (xorima) [inline resources/server_install.rb]: We don't want to use attributes inside here, they are plain text on the server, please revert to a sane default ie `'generate'`

File: resources/user.rb [hunk 3]
```diff
@@ -72,7 +72,7 @@ def run_query(query)
 
   def database_has_password_column
     begin
-      result = run_query('SHOW COLUMNS FROM mysql.user WHERE Field="Password"')
+      result = run_query("SHOW COLUMNS FROM mysql.user WHERE Field='Password';")
```
Comment (xorima) [inline resources/user.rb:75]: Why the change here?

**Review by xorima — CHANGES_REQUESTED** [review]

File: resources/user.rb (hunk 3 above)
Comment (t0rrant) [reply resources/user.rb:75]: This setup of quotes is what causes #220, the query is formed to:

```
"SHOW COLUMNS FROM mysql.user WHERE Field="Password""
```

which results in a non-valid query

File: resources/server_install.rb (hunk 2 above)
Comment (t0rrant) [reply resources/server_install.rb]: ok I'll review the password generation code and see what is affected by that.

File: metadata.rb (hunk 1 above)
Comment (t0rrant) [reply metadata.rb]: sure thing =)

Comment (t0rrant) [reply metadata.rb]: how about the changelog?

File: resources/server_install.rb (hunk 2 above)
Comment (t0rrant) [reply resources/server_install.rb]: everything should stay the same with that revert, however, I used `lazy { node['mariadb']['server_root_password'] }` as the default due to this misleading [comment](https://github.com/sous-chefs/mariadb/blob/master/README.md#syntax)

File: metadata.rb (hunk 1 above)
Comment (t0rrant) [reply metadata.rb]: In the last commit I left the version reference untouched in the CHANGELOG.md with the changes made.

File: resources/server_install.rb (hunk 2 above)
Comment (t0rrant) [reply resources/server_install.rb]: I reverted that and removed from CHANGELOG.md indication about the `server_root_password` usage for the default value.

**Comment by johnbellone** [comment]
I am running into this issue. It'd be awesome if we could fix it!

**Comment by t0rrant** [comment]
@johnbellone you can use my fork until they accept and merge the PR.

You can do that by adding this to your Berksfile:

```cookbook 'mariadb', '~> 2', git: 'https://github.com/t0rrant/mariadb.git'```

Let me know if you come across any other issues.

File: resources/server_install.rb [hunk 4]
```diff
@@ -49,21 +49,59 @@
...
+  #                                             2- the user did not pass anything to the password argument OR
+  #                                                the user did not define node['mariadb']['server_root_password'] attribute
+  mariadb_root_password = (new_resource.password == 'generate' || new_resource.password.nil?) ? secure_random : new_resource.password
 
-  # Generate a ramdom password or set the a password defined with node['mariadb']['password']['root'].
+  # Generate a random password or set a password defined with node['mariadb']['server_root_password'].
   # The password is set or change at each run. It is good for security if you choose to set a random password and
   # allow you to change the root password if needed.
-  bash 'generate-mariadb-root-password' do
+  file 'generate-mariadb-root-password' do
```
Comment (xorima) [inline resources/server_install.rb:64]: Do we really want to be putting the password in plain text on the server?

Comment (t0rrant) [reply resources/server_install.rb:64]: I see your point. However, the file is only accessible by root and mysql users, and honestly if the machine is somehow owned, the attacker could always access the db server anyway.

In the end the created file can be removed, but (to my knowledge) it has to exist, keep in mind the `--init-file=` argument needed to pass to mysqld to reset the root password.

Of course other solutions are welcome.

File: resources/client_install.rb [hunk 5]
```diff
@@ -22,7 +22,7 @@
 action :install do
   mariadb_repository 'Add mariadb.org repository' do
     version new_resource.version
-    only_if { new_resource.setup_repo }
+    only_if { !new_resource.setup_repo.nil? && new_resource.setup_repo }
```
Comment (shoekstra) [inline resources/client_install.rb]: By saying `new_resource.setup_repo` you are implying it is not `nil` so you don't need to say `new_resource.setup_repo`; this change can be reverted.

File: CHANGELOG.md [hunk 6]
```diff
@@ -4,6 +4,33 @@ All notable changes to this project will be documented in this file.
...
+- Property `apt_key_proxy` to `mariadb_repository` to be able to pass a proxy setting to apt-key ([#234](https://github.com/sous-chefs/mariadb/pull/234/commits/d5b09122492d5474b38ad00a454bf49175a12c79))
+- A mechanism to properly set the root password for mariadb ([#234-bc33fb2](https://github.com/sous-chefs/mariadb/pull/234/commits/bc33fb2c9dfc0a14754eb76bbff43ac6c3346d5b) :
+[#234-c576d42](https://github.com/sous-chefs/mariadb/pull/234/commits/c576d42a7317d63d87a0cff44ad7c564a9b1a0e2))
+
+  
+### Fixed
+
+- String quoting that prevented some SQL commands to execute properly ([#220](https://github.com/sous-chefs/mariadb/issues/220))
+- Changelog broken links
+
```
Comment (shoekstra) [inline CHANGELOG.md:21]: Hi!  Could all these be bundled up under a "Unreleased" version?

File: test/cookbooks/test/recipes/server_install.rb [hunk 7]
```diff
@@ -9,6 +9,6 @@
 find_resource(:service, 'mariadb') do
   extend MariaDBCookbook::Helpers
   service_name lazy { platform_service_name }
-  supports restart: true, status: true, reload: true
+  supports restart: true, status: true
```
Comment (shoekstra) [inline test/cookbooks/test/recipes/server_install.rb]: The service supports `reload` so this should be reverted.

File: resources/server_install.rb [hunk 8]
```diff
@@ -49,21 +49,59 @@
...
+  execute 'verify-root-password-okay' do
     user 'root'
-    code <<-EOH
-    echo "ALTER USER 'root'@'localhost' IDENTIFIED BY \'#{mariadb_root_password}\';" | /usr/bin/mysql
-    EOH
-    not_if { ::File.exist? "#{data_dir}/recovery.conf" }
-    only_if { new_resource.password }
+    command "mysql -u root -p#{mariadb_root_password} -e '\\s'&>/dev/null"
+    action :nothing
   end
```
Comment (shoekstra) [inline resources/server_install.rb:97]: I'm not sure why we keep this - we should validate the way Chef configures the system by using integration tests, not by calling another resource in the recipe.

Would you be open to removing this and creating an InSpec test instead?

File: resources/server_install.rb [hunk 9]
```diff
@@ -49,21 +49,59 @@
...
+    mode '640'
+    sensitive true
+    content "use mysql;
+update user set password=PASSWORD('#{mariadb_root_password}') where User='root';
+flush privileges;"
+    action :nothing
+  end
+
+  pid_file = default_pid_file.nil? ? '/var/run/mysqld/mysqld.pid' : default_pid_file
+  pid_dir = ::File.dirname(pid_file).to_s
```
Comment (shoekstra) [inline resources/server_install.rb]: I don't think you need the `.to_s` here.

File: resources/server_install.rb [hunk 10]
```diff
@@ -49,21 +49,59 @@
...
+
+  # because some distros may not take care of the pid file location directory, we manage it ourselves
+  directory pid_dir.to_s do
```
Comment (shoekstra) [inline resources/server_install.rb]: Same here re `.to_s`

**Review by shoekstra — CHANGES_REQUESTED** [review]

**Comment by shoekstra** [comment]
Hi!  I've added some feedback for you to review.

For future it's good to keep fixes/features in separate PRs as it makes it easier to review and quicker to get merged in.

File: CHANGELOG.md (hunk 6 above)
Comment (t0rrant) [reply CHANGELOG.md:21]: Sure, will do that along with the rest of the changes.

File: resources/client_install.rb (hunk 5 above)
Comment (t0rrant) [reply resources/client_install.rb]: indeed, it's already true or false, never nil.

**Comment by t0rrant** [comment]
Hey @shoekstra,

I'll keep that in mind for future PRs, this kind of blew up while trying to solve one thing.

Sorry for the mess.

File: resources/server_install.rb (hunk 8 above)
Comment (t0rrant) [reply resources/server_install.rb:97]: I agree with you. However, I don't think it's possible to access from InSpec the password value we use here.

Comment (t0rrant) [reply resources/server_install.rb:97]: I'll remove that validation, as it can also pose a security issue and it really adds nothing to the resource.

**Comment by kitchen-porter** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This Pull Request is probably missing tests.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by t0rrant** [comment]
rubocop update? lint broken unrelated to this PR.

**Comment by aspyatkin** [comment]
Hello, any idea when this gets released? Seems to be a blocker

**Review by shoekstra — APPROVED** [review]

**Review by xorima — APPROVED** [review]

**Pull request merged**